### DISCUSSION TITLE
UseEntry does not recognize type: asset/resource

### DISCUSSION THREAD

**Discussion by tranvansang** [webpack/webpack#14703] [General]
# Bug report

**What is the current behavior?**

The following error occurs, and webpack cannot compile.

```
[webpack-cli] Invalid configuration object. Webpack has been initialized using a configuration object that does not match the API schema.
 - configuration.module.rules[0].use[0] has an unknown property 'type'. These properties are valid:
   object { ident?, loader?, options? }
```

**If the current behavior is a bug, please provide the steps to reproduce.**

- Clone this repo https://github.com/tranvansang/webpack-rule-bug
- `yarn`
- `yarn start` ==> failed
- Open `webpack.config.js`, uncomment line 15, and comment out line 16.

Before:
```
					{
						//loader: 'url-loader' // worked
						type: 'asset/resource', // not work
					},

```

After:
```
					{
						loader: 'url-loader' // worked
						//type: 'asset/resource', // not work
					},

```
- `yarn start` again ==> success

**What is the expected behavior?**

UseEntry should recognize the `type: 'asset/resource'` entry.
Currently, the only way I can achieve loading my resource in nested UseEntry is to use `url-loader`.
However, `url-loader` is deprecated.

**Other relevant information:**
webpack version:  5.63.0
Node.js version: v16.5.0
Operating System: 5.14.10-arch1-1 x86_64 GNU/Linux
Additional tools: webpack-cli v4.9.1, url-loader v 4.1.1

**Comment by alexander-akait** [comment]
Please provide configuration

**Reply by tranvansang** [reply]
https://github.com/tranvansang/webpack-rule-bug/blob/1f4822e8d89339ad7c7632e211c2a8b3c50c4dd6/webpack.config.js

This is the configuration.

```javascript
module.exports = {
	entry: './in.js',
	output: {
		filename: 'out.js',
		path: __dirname,
	},
	module: {
		rules: [
			{
				test: /\.png/,
				use: [
					{
						//loader: 'url-loader' // worked
						type: 'asset/resource', // not work
					},
				]
			}
		]
	}
}
```

**Comment by alexander-akait** [comment]
Move `type` on `test` level, type can be nested

**Reply by tranvansang** [reply]
I have another loader, it converts the file to a buffer and I want to pipeline its output buffer to the `asset/resource`.

Unfortunately, your solution does not work in my case.

**Reply by alexander-akait** [reply]
`type` is not loader, `type` is module definition, you can't set it on `loader` level

**Reply by tranvansang** [reply]
so if I want to chain my buffer to an asset, I have to use `url-loader`?

Does this mean that `url-loader` shouldn't be deprecated until `asset/resource` is supported?

**Reply by alexander-akait** [reply]
Why do not move `type` on level above, you don't need it nested, `type` will be handled after all loaders, if you provide more details I will show you how to solve it

**Reply by tranvansang** [reply]
I want to change file width before exporting it to assets.

`import a from 'a.png?width=500`

I also have another custom loader by testing the query, like

`import a from 'a.png?size'`

My current loader is

```javascript
{
  test: /\.png/,
  oneOf: [
    {resourceQuery: /size/, use: ['size-loader']},
    {resourceQuery: /width/, use: ['file-loader', 'image-convert-loader']},
    {type: 'asset/resource'}
  ]
}
```

Because 'file-loader' is deprecated, I'd like to replace it with 'asset/resource'.

Do you have any idea? Thanks very much.

**Reply by alexander-akait** [reply]
Just remove `file-loader` and set `type: 'asset/resource'` on `test` level, another solution avoid to use `oneOf`, if you provide small reproducible example I will show how to do it

**Reply by tranvansang** [reply]
it works like a charm. Thank you very much!!

I didn't notice that `type` can placed together with `loader`.
